src/tool/hpcrun/cct/cct.c

```diff
@@ -222,28 +225,62 @@ splay(cct_node_t* cct, cct_addr_t* addr)
 // helper for walking functions
 //
 
+// Get the leftmost child of a splay node, marking splay-parents while
+// descending.
 //
-// lrs abbreviation for "left-right-self"
+// Do not use directly, use splay_walk_* instead.
+static cct_node_t* splay_leftmost_descent(cct_node_t* cur) {
+  while(cur->left != NULL || cur->right != NULL) {
+    cct_node_t* next = cur->left != NULL ? cur->left : cur->right;
+    next->up = cur;
+    cur = next;
+  }
+  return cur;
+}
+
+// Get the first node in a post-order traversal of the splay tree of sibling
+// cct nodes.
 //
-static void
-walk_child_lrs(cct_node_t* cct,
-               cct_op_t op, cct_op_arg_t arg, size_t level,
-               void (*wf)(cct_node_t* n, cct_op_t o, cct_op_arg_t a, size_t l))
-{
-  if (!cct) return;
+// If there are no nodes to walk (ie. the argument is NULL), returns NULL.
+//
+// See splay_walk_next for more details.
+static cct_node_t* splay_walk_init(cct_node_t* root) {
+  if(root == NULL) return NULL;
+  assert(root->up == NULL);
 
-  walk_child_lrs(cct->left, op, arg, level, wf);
-  walk_child_lrs(cct->right, op, arg, level, wf);
-  wf(cct, op, arg, level);
+  // First node in the iteration is always the leftmost child.
+  return splay_leftmost_descent(root);
```

Comment:
why splaying at each level of the walk?

I'm going to need to discuss the rationale for this with @blue42u

Reply:
`splay` as in "splay tree," this function doesn't alter the tree's shape, it only fiddles with the `up` field.

Specifically, `spack_leftmost_descent` returns the leftmost lowest child of the splay tree rooted at `root`, setting the `up` fields to allow for travel back up the tree when needed.